sycl/include/CL/sycl/detail/kernel_program_cache.hpp

```diff
@@ -0,0 +1,92 @@
+//==--- kernel_program_cache.hpp - Cache for kernel and program -*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/locked.hpp>
+#include <CL/sycl/detail/os_util.hpp>
+#include <CL/sycl/detail/pi.hpp>
+
+#include <atomic>
+#include <condition_variable>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <type_traits>
+
+namespace cl {
+namespace sycl {
+namespace detail {
+class kernel_program_cache {
+public:
+  using PiProgramT = std::remove_pointer<RT::PiProgram>::type;
+  using PiProgramPtrT = std::atomic<PiProgramT *>;
+
+  template<typename T>
+  struct SmthWithBuildState {
```

Comment:
Feel free to comment a lot the .hpp... :-)
For example I have no idea about what is an `Smth` here...